File: Engine/Android/core/src/test/kotlin/despia/engine/CalendarConformanceTest.kt
package despia.engine

import java.io.File
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * The calendar conformance runner - executes
 * OpenSource/Conformance/calendar/{crud,present,recurrence}.json through THIS runtime's
 * CalendarCore folds (parity F12). The TS twin (@despia/kernel calendar-core.ts,
 * calendar-conformance.test.ts) and the Swift reference (CalendarConformance) run the SAME
 * files, so the `futureEvents` blast radius, the iOS 17 writeOnly split, the editor
 * result-fidelity ladder and an RRULE round-trip cannot drift between renderers.
 *
 * Missing corpus = loud failure, and so is an empty section.
 */
class CalendarConformanceTest {

    private fun corpusFile(name: String): File {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/calendar/$name")
            if (candidate.isFile) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/calendar/$name not found")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun root(name: String): Map<String, Any?> {
        val root = json(corpusFile(name).readText()).foundationValue as? Map<String, Any?>
            ?: error("$name: not a JSON object")
        assertEquals(1, (root["version"] as? Number)?.toInt(), "$name version")
        return root
    }

    @Suppress("UNCHECKED_CAST")
    private fun section(root: Map<String, Any?>, name: String, file: String): List<Map<String, Any?>> {
        val raw = root[name]
        val rows = (raw as? List<Any?>) ?: ((raw as? Map<String, Any?>)?.get("cases") as? List<Any?>)
        val cases = rows?.map { it as Map<String, Any?> }
        assertTrue(cases != null && cases.isNotEmpty(), "$file: $name must be a non-empty case array")
        return cases!!
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun spanDecisionAgreesWithCorpus() {
        for (case in section(root("crud.json"), "span", "crud.json")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val given = case["given"] as Map<String, Any?>
            assertEquals(
                (case["expect"] as Map<String, Any?>)["scope"],
                CalendarCore.span(given["recurring"] == true, given["futureEvents"] as? Boolean),
                "$name: scope",
            )
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun dateGrammarAgreesWithCorpus() {
        for (case in section(root("crud.json"), "dates", "crud.json")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val input = case["input"]
            val expect = case["expect"] as Map<String, Any?>

            if (input is Map<*, *>) {
                val window = CalendarCore.window(input["start"], input["end"])
                assertEquals(expect["error"] as String?, if (window == null) "invalid_date" else null,
                             "$name: window")
                continue
            }
            val epoch = CalendarCore.parseDate(input)
            val expectedError = expect["error"] as? String
            if (expectedError != null) {
                assertNull(epoch, "$name: expected $expectedError")
                continue
            }
            assertEquals((expect["epoch"] as Number).toDouble(), epoch, "$name: epoch")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun accessSplitAgreesWithCorpus() {
        for (case in section(root("crud.json"), "access", "crud.json")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val given = case["given"] as Map<String, Any?>
            val expect = case["expect"] as Map<String, Any?>
            val decision = CalendarCore.accessDecision(given["access"] as? String, given["action"] as String)

            assertEquals(expect["runs"], decision.runs, "$name: runs")
            if (expect.containsKey("prompted")) {
                assertEquals(expect["prompted"], decision.prompted, "$name: prompted")
            }
            assertEquals(expect["error"] as String?, decision.error, "$name: error")
            if (expect.containsKey("messageNames")) {
                val names = (expect["messageNames"] as String).lowercase()
                assertTrue(decision.message.orEmpty().lowercase().contains(names),
                           "$name: the refusal must name $names, got ${decision.message}")
            }
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun calendarTargetsAgreeWithCorpus() {
        for (case in section(root("crud.json"), "calendars", "crud.json")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val given = case["given"] as Map<String, Any?>
            val expect = case["expect"] as Map<String, Any?>
            val decision = CalendarCore.targetDecision(
                calendarId = given["calendarId"] as? String,
                exists = given["exists"] as? Boolean,
                writable = given["writable"] as? Boolean,
                hasDefault = given["hasDefault"] as? Boolean,
            )
            assertEquals(expect["runs"], decision.runs, "$name: runs")
            assertEquals(expect["error"] as String?, decision.error, "$name: error")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun remindersAbsenceAgreesWithCorpus() {
        for (case in section(root("crud.json"), "reminders", "crud.json")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val given = case["given"] as Map<String, Any?>
            val expect = case["expect"] as Map<String, Any?>
            val support = CalendarCore.remindersSupport(given["renderer"] as String)
            assertEquals(expect["supported"], support.supported, "$name: supported")
            assertEquals(expect["error"] as String?, support.error, "$name: error")
            if (expect.containsKey("remindersAccess")) {
                assertEquals(expect["remindersAccess"] as String?, support.remindersAccess,
                             "$name: remindersAccess")
            }
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun presentPermissionSurfaceAgreesWithCorpus() {
        val surface = root("present.json")["permissionSurface"] as Map<String, Any?>
        val declared = surface.filterKeys { !it.startsWith("_") }
        assertTrue(declared.isNotEmpty(), "present.json: permissionSurface names no actions")
        for ((action, grant) in declared) {
            assertEquals(grant, CalendarCore.PERMISSION_SURFACE[action], "permissionSurface: $action")
        }
        for (action in CalendarCore.PERMISSION_SURFACE.keys) {
            assertTrue(surface.containsKey(action), "permissionSurface: the corpus does not pin $action")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun resultFidelityAgreesWithCorpus() {
        val fidelity = root("present.json")["resultFidelity"] as Map<String, Any?>
        val declared = fidelity.filterKeys { !it.startsWith("_") }
        assertTrue(declared.isNotEmpty(), "present.json: resultFidelity names no renderers")
        for ((renderer, results) in declared) {
            assertEquals((results as List<Any?>).map { it as String },
                         CalendarCore.RESULT_FIDELITY[renderer], "resultFidelity: $renderer")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun presentOutcomeAgreesWithCorpus() {
        val doc = root("present.json")
        val fidelity = doc["resultFidelity"] as Map<String, Any?>
        for (case in section(doc, "cases", "present.json")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val given = case["given"] as Map<String, Any?>
            val expect = case["expect"] as Map<String, Any?>

            val outcome = CalendarCore.presentOutcome(
                access = given["access"] as? String,
                editorAction = given["editorAction"] as? String,
                eventFound = given["eventFound"] as? Boolean,
                id = given["id"] as? String,
                start = given["start"],
                end = given["end"],
                hasWindow = given.containsKey("start") || given.containsKey("end"),
            )

            val expectedError = expect["error"] as? String
            if (expectedError != null) {
                assertEquals(expectedError, outcome.error, "$name: error")
                assertEquals(expect["presented"] ?: false, outcome.presented, "$name: presented")
                continue
            }
            assertNull(outcome.error, "$name: error")
            assertEquals(expect["result"], outcome.result, "$name: result")
            if (expect.containsKey("hasId")) assertEquals(expect["hasId"], outcome.hasId, "$name: hasId")
            if (expect.containsKey("broadcast")) {
                assertEquals(expect["broadcast"] as String?, outcome.broadcast, "$name: broadcast")
            }
            val allowed = (fidelity[given["renderer"] as String] as List<Any?>).map { it as String }
            assertTrue(outcome.result in allowed,
                       "$name: ${outcome.result} is outside ${given["renderer"]}'s fidelity list")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun recurrenceRoundTripAgreesWithCorpus() {
        for (case in section(root("recurrence.json"), "roundTrip", "recurrence.json")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val rule = CalendarCore.parseRecurrence(case["rrule"] as String)
            assertNotNull(rule, "$name: expected a parsed rule")

            val parsed = case["parsed"] as Map<String, Any?>
            assertEquals(parsed["freq"], rule.freq, "$name: freq")
            assertEquals((parsed["interval"] as Number).toInt(), rule.interval, "$name: interval")
            if (parsed.containsKey("count")) {
                assertEquals((parsed["count"] as Number).toInt(), rule.count, "$name: count")
            }
            if (parsed.containsKey("until")) {
                assertEquals(parsed["until"], rule.until, "$name: until")
            }
            if (parsed.containsKey("byDay")) {
                val expected = (parsed["byDay"] as List<Any?>).map {
                    val row = it as Map<String, Any?>
                    RecurrenceDay(row["day"] as String, (row["ordinal"] as Number).toInt())
                }
                assertEquals(expected, rule.byDay, "$name: byDay")
            }
            if (parsed.containsKey("byMonthDay")) {
                assertEquals((parsed["byMonthDay"] as List<Any?>).map { (it as Number).toInt() },
                             rule.byMonthDay, "$name: byMonthDay")
            }
            if (parsed.containsKey("byMonth")) {
                assertEquals((parsed["byMonth"] as List<Any?>).map { (it as Number).toInt() },
                             rule.byMonth, "$name: byMonth")
            }
            assertEquals(case["canonical"], CalendarCore.formatRecurrence(rule), "$name: canonical")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun rejectedRecurrenceIsRefusedLoudly() {
        for (case in section(root("recurrence.json"), "rejected", "recurrence.json")) {
            val name = case["name"] as? String ?: "<unnamed>"
            assertNull(CalendarCore.parseRecurrence(case["rrule"] as String), "$name: must not parse")
            assertEquals("invalid_recurrence", case["error"], "$name: error code")
        }
    }
}

File: Web/packages/kernel/test/calendar-conformance.test.ts
//
//  calendar-conformance.test.ts — the SHARED calendar corpus through the TS kernel
//  (OpenSource/Conformance/calendar/{crud,present,recurrence}.json). The Kotlin twin (:core
//  CalendarConformanceTest) and the Swift reference (CalendarConformance, record lane) run the
//  SAME files, so the `futureEvents` blast radius, the iOS 17 writeOnly split, the editor
//  result-fidelity ladder and an RRULE round-trip cannot mean one thing on one renderer and
//  something else on another.
//
//  Missing corpus = loud failure, and so is an empty section.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import {
  calendarSpan, parseCalendarDate, calendarWindow, calendarAccessDecision,
  calendarTargetDecision, calendarRemindersSupport, calendarPresentOutcome,
  parseRecurrence, formatRecurrence, CALENDAR_PERMISSION_SURFACE, CALENDAR_RESULT_FIDELITY,
} from "../src/calendar-core.ts";

type Json = { [key: string]: any };

function corpusDir(): string {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, "OpenSource/Conformance/calendar");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("OpenSource/Conformance/calendar not found");
    dir = parent;
  }
}

function corpus(file: string): Json {
  const doc = JSON.parse(readFileSync(join(corpusDir(), file), "utf-8")) as Json;
  assert.equal(doc["version"], 1, `${file}: version`);
  return doc;
}

function section(doc: Json, name: string, file: string): Json[] {
  const raw = doc[name];
  const rows = Array.isArray(raw) ? raw : (raw as Json | undefined)?.["cases"];
  assert.ok(Array.isArray(rows) && rows.length > 0, `${file}: ${name} must be a non-empty case array`);
  return rows as Json[];
}

test("calendar: the futureEvents span decision agrees with the corpus", () => {
  for (const testCase of section(corpus("crud.json"), "span", "crud.json")) {
    const given = testCase["given"] as Json;
    assert.equal(
      calendarSpan(given["recurring"] === true, given["futureEvents"] as boolean | null),
      (testCase["expect"] as Json)["scope"],
      `${testCase["name"]}: scope`,
    );
  }
});

test("calendar: the date grammar agrees with the corpus", () => {
  for (const testCase of section(corpus("crud.json"), "dates", "crud.json")) {
    const input = testCase["input"];
    const expect = testCase["expect"] as Json;
    const name = testCase["name"] as string;

    if (input !== null && typeof input === "object" && !Array.isArray(input)) {
      const window = calendarWindow((input as Json)["start"], (input as Json)["end"]);
      const error = window === null ? "invalid_date" : null;
      assert.equal(error, expect["error"] ?? null, `${name}: window`);
      continue;
    }
    const epoch = parseCalendarDate(input);
    if (expect["error"] !== undefined && expect["error"] !== null) {
      assert.equal(epoch, null, `${name}: expected ${expect["error"]}`);
      continue;
    }
    assert.equal(epoch, expect["epoch"], `${name}: epoch`);
  }
});

test("calendar: the access split agrees with the corpus", () => {
  for (const testCase of section(corpus("crud.json"), "access", "crud.json")) {
    const given = testCase["given"] as Json;
    const expect = testCase["expect"] as Json;
    const name = testCase["name"] as string;
    const decision = calendarAccessDecision(given["access"] as string, given["action"] as string);

    assert.equal(decision.runs, expect["runs"], `${name}: runs`);
    if (expect["prompted"] !== undefined) assert.equal(decision.prompted, expect["prompted"], `${name}: prompted`);
    assert.equal(decision.error, expect["error"] ?? null, `${name}: error`);
    if (expect["messageNames"] !== undefined) {
      assert.ok(
        String(decision.message ?? "").toLowerCase().includes(String(expect["messageNames"]).toLowerCase()),
        `${name}: the refusal must name ${expect["messageNames"]}, got ${decision.message}`,
      );
    }
  }
});

test("calendar: the calendar-target refusals agree with the corpus", () => {
  for (const testCase of section(corpus("crud.json"), "calendars", "crud.json")) {
    const given = testCase["given"] as Json;
    const expect = testCase["expect"] as Json;
    const decision = calendarTargetDecision({
      calendarId: given["calendarId"] as string | null,
      exists: given["exists"] as boolean | undefined,
      writable: given["writable"] as boolean | undefined,
      hasDefault: given["hasDefault"] as boolean | undefined,
    });
    assert.equal(decision.runs, expect["runs"], `${testCase["name"]}: runs`);
    assert.equal(decision.error, expect["error"] ?? null, `${testCase["name"]}: error`);
  }
});

test("calendar: the reminders absence agrees with the corpus", () => {
  for (const testCase of section(corpus("crud.json"), "reminders", "crud.json")) {
    const given = testCase["given"] as Json;
    const expect = testCase["expect"] as Json;
    const support = calendarRemindersSupport(given["renderer"] as string);
    assert.equal(support.supported, expect["supported"], `${testCase["name"]}: supported`);
    assert.equal(support.error, expect["error"] ?? null, `${testCase["name"]}: error`);
    if (expect["remindersAccess"] !== undefined) {
      assert.equal(support.remindersAccess, expect["remindersAccess"], `${testCase["name"]}: remindersAccess`);
    }
  }
});

test("calendar: the present permission surface agrees with the corpus", () => {
  const surface = corpus("present.json")["permissionSurface"] as Json;
  const declared = Object.entries(surface).filter(([key]) => !key.startsWith("_"));
  assert.ok(declared.length > 0, "present.json: permissionSurface names no actions");
  for (const [action, grant] of declared) {
    assert.equal(CALENDAR_PERMISSION_SURFACE[action], grant, `permissionSurface: ${action}`);
  }
  for (const action of Object.keys(CALENDAR_PERMISSION_SURFACE)) {
    assert.ok(action in surface, `permissionSurface: the corpus does not pin ${action}`);
  }
});

test("calendar: the result-fidelity table agrees with the corpus", () => {
  const fidelity = corpus("present.json")["resultFidelity"] as Json;
  const declared = Object.entries(fidelity).filter(([key]) => !key.startsWith("_"));
  assert.ok(declared.length > 0, "present.json: resultFidelity names no renderers");
  for (const [renderer, results] of declared) {
    assert.deepEqual([...(CALENDAR_RESULT_FIDELITY[renderer] ?? [])], results, `resultFidelity: ${renderer}`);
  }
});

test("calendar: the present outcome ladder agrees with the corpus", () => {
  const doc = corpus("present.json");
  const fidelity = doc["resultFidelity"] as Json;
  for (const testCase of section(doc, "cases", "present.json")) {
    const given = testCase["given"] as Json;
    const expect = testCase["expect"] as Json;
    const name = testCase["name"] as string;

    const outcome = calendarPresentOutcome({
      renderer: given["renderer"] as string,
      access: given["access"] as string,
      editorAction: given["editorAction"] as string | null,
      eventFound: given["eventFound"] as boolean | undefined,
      id: given["id"] as string | undefined,
      start: given["start"],
      end: given["end"],
    });

    if (expect["error"] !== undefined && expect["error"] !== null) {
      assert.equal(outcome.error, expect["error"], `${name}: error`);
      assert.equal(outcome.presented, expect["presented"] ?? false, `${name}: presented`);
      continue;
    }
    assert.equal(outcome.error, null, `${name}: error`);
    assert.equal(outcome.result, expect["result"], `${name}: result`);
    if (expect["hasId"] !== undefined) assert.equal(outcome.hasId, expect["hasId"], `${name}: hasId`);
    if ("broadcast" in expect) assert.equal(outcome.broadcast, expect["broadcast"], `${name}: broadcast`);
    assert.ok(
      (fidelity[given["renderer"] as string] as string[]).includes(outcome.result as string),
      `${name}: ${outcome.result} is outside ${given["renderer"]}'s fidelity list`,
    );
  }
});

test("calendar: the RRULE round-trip agrees with the corpus", () => {
  for (const testCase of section(corpus("recurrence.json"), "roundTrip", "recurrence.json")) {
    const name = testCase["name"] as string;
    const rule = parseRecurrence(testCase["rrule"] as string);
    assert.ok(rule !== null, `${name}: expected a parsed rule`);

    const parsed = testCase["parsed"] as Json;
    assert.equal(rule.freq, parsed["freq"], `${name}: freq`);
    assert.equal(rule.interval, parsed["interval"], `${name}: interval`);
    if (parsed["count"] !== undefined) assert.equal(rule.count, parsed["count"], `${name}: count`);
    if (parsed["until"] !== undefined) assert.equal(rule.until, parsed["until"], `${name}: until`);
    if (parsed["byDay"] !== undefined) {
      assert.deepEqual(rule.byDay?.map((d) => ({ day: d.day, ordinal: d.ordinal })), parsed["byDay"], `${name}: byDay`);
    }
    if (parsed["byMonthDay"] !== undefined) {
      assert.deepEqual([...(rule.byMonthDay ?? [])], parsed["byMonthDay"], `${name}: byMonthDay`);
    }
    if (parsed["byMonth"] !== undefined) {
      assert.deepEqual([...(rule.byMonth ?? [])], parsed["byMonth"], `${name}: byMonth`);
    }
    assert.equal(formatRecurrence(rule), testCase["canonical"], `${name}: canonical`);
  }
});

test("calendar: every rejected RRULE is refused loudly", () => {
  for (const testCase of section(corpus("recurrence.json"), "rejected", "recurrence.json")) {
    assert.equal(parseRecurrence(testCase["rrule"] as string), null, `${testCase["name"]}: must not parse`);
    assert.equal(testCase["error"], "invalid_recurrence", `${testCase["name"]}: error code`);
  }
});
